gpiozero/devices.py

```diff
@@ -229,13 +189,100 @@ def __exit__(self, exc_type, exc_value, exc_tb):
 class Device(ValuesMixin, GPIOBase):
     """
     Represents a single device of any type; GPIO-based, SPI-based, I2C-based,
-    etc. This is the base class of the device hierarchy. It defines the
-    basic services applicable to all devices (specifically the :attr:`is_active`
+    etc. This is the base class of the device hierarchy. It defines the basic
+    services applicable to all devices (specifically the :attr:`is_active`
     property, the :attr:`value` property, and the :meth:`close` method).
     """
+    _pin_factory = None # instance of a Factory sub-class
+    _reservations = {} # maps pin addresses to lists of devices
+    _res_lock = Lock()
+
     def __repr__(self):
         return "<gpiozero.%s object>" % (self.__class__.__name__)
 
+    @classmethod
+    def _set_pin_factory(cls, new_factory):
+        if cls._pin_factory is not None:
+            cls._pin_factory.close()
+        cls._pin_factory = new_factory
+
+    def _reserve_pins(self, *pins_or_addresses):
+        """
+        Called to indicate that the device reserves the right to use the
+        specified *pins_or_addresses*. This should be done during device
+        construction.  If pins are reserved, you must ensure that the
+        reservation is released by eventually called :meth:`_release_pins`.
+
+        The *pins_or_addresses* can be actual :class:`Pin` instances or the
+        addresses of pin instances (each address is a tuple of strings). The
+        latter form is permitted to ensure that devices do not have to
+        construct :class:`Pin` objects to reserve pins. This is important as
+        constructing a pin often configures it (e.g. as an input) which
+        conflicts with alternate pin functions like SPI.
+        """
+        addresses = (
+            p.address if isinstance(p, Pin) else p
+            for p in pins_or_addresses
+            )
+        with self._res_lock:
+            for address in addresses:
+                try:
+                    conflictors = self._reservations[address]
+                except KeyError:
+                    conflictors = []
+                    self._reservations[address] = conflictors
+                for device_ref in conflictors:
+                    device = device_ref()
+                    if device is not None and self._conflicts_with(device):
+                        raise GPIOPinInUse(
+                            'pin %s is already in use by %r' % (
+                                '/'.join(address), device)
+                        )
+                conflictors.append(weakref.ref(self))
```

Comment:
IMHO this looks a bit odd - seems like you're only modifying a local variable.
I realise the code is actually equivalent (object references etc.), but I wonder if
```python
self._reservations[address].append(weakref.ref(self))
```
would make it clearer that this method *is* actually modifying instance-state. (IYSWIM?)